packages/pages/src/dev/server/server.ts

```diff
@@ -61,22 +62,41 @@ export const createServer = async (
     ));
   }
 
-  const loadedFunctions = await loadServerlessFunctions();
-  const loadedFunctionPaths = [...loadedFunctions.keys()].map(
-    (slug) => "/" + slug
-  );
+  // Load functions from their source files
+  const functionModules: Map<string, FunctionModuleInternal> = new Map();
+  const loadUpdatedFunctionModules = async () => {
+    const loadedFunctionModules = await loadFunctions();
+    loadedFunctionModules.forEach((functionModule) => {
+      functionModules.set(functionModule.slug, functionModule);
+    });
+  };
+
+  await loadUpdatedFunctionModules(); // Load functions on initial setup
 
-  if (loadedFunctionPaths.length > 0) {
-    loadedFunctionPaths.forEach((loadedFunctionPath) => {
-      const loadedFunction = loadedFunctions.get(loadedFunctionPath.slice(1));
-      if (loadedFunction) {
-        app.use(loadedFunctionPath, (req, res) =>
-          serveServerlessFunction(req, res, loadedFunction)
-        );
-      }
+  // Assign routes for functions based on their slug when the server started
+  const loadedFunctionSlugs = [...functionModules.keys()];
+  if (loadedFunctionSlugs.length > 0) {
+    loadedFunctionSlugs.forEach((loadedFunctionSlug) => {
+      app.use("/" + loadedFunctionSlug, (req, res, next) => {
+        const loadedFunction = functionModules.get(loadedFunctionSlug);
+        if (!loadedFunction) {
+          throw new Error(
+            "Could not load function with slug" + loadedFunctionSlug
+          );
+        }
+        serveServerlessFunction(req, res, next, loadedFunction);
+      });
     });
   }
 
+  // Reload functions if their source files change
+  // Only supports updates in the default export, changes to the slug require server restart
+  vite.watcher.on("change", async (filepath) => {
```

Comment:
Does HMR not work for these out of the box? If you refresh the function page does the updated code show up?

Reply:
Yeah it doesn't work automatically for the functions. Vite registers a reload but then doesn't load the new code or something -- I'm not familiar with HMR so I'm not really sure why. I can dive into it more if you want

Reply:
If you refresh the page after saving your code change does it work without this? And with this what happens? The page will refresh after you save?